File: list-sort/haskell/Sort.hs
{-# LANGUAGE BangPatterns #-}
-- Sorting benchmark: quicksort and mergesort on Word64 arrays (mutable IOVector)
-- LCG PRNG (Knuth) — identical across all languages

module Main where

import Data.Word (Word64)
import qualified Data.Vector.Unboxed.Mutable as MV
import qualified Data.Vector.Unboxed as V
import System.Environment (getArgs)
import System.Clock (getTime, Clock(Monotonic), toNanoSecs)

-- LCG PRNG
lcgNext :: Word64 -> (Word64, Word64)
lcgNext s = let s' = s * 6364136223846793005 + 1442695040888963407 in (s', s')

generateArray :: Int -> Word64 -> IO (MV.IOVector Word64)
generateArray n seed = do
    arr <- MV.new n
    let go !i !s
          | i >= n    = return arr
          | otherwise = let (!val, !s') = lcgNext s
                        in MV.write arr i val >> go (i+1) s'
    go 0 seed

-- Lomuto partition
partition :: MV.IOVector Word64 -> Int -> Int -> IO Int
partition arr lo hi = do
    pivot <- MV.read arr hi
    let go !j !i
          | j >= hi   = MV.swap arr i hi >> return i
          | otherwise = do
              v <- MV.read arr j
              if v <= pivot
                then MV.swap arr i j >> go (j+1) (i+1)
                else go (j+1) i
    go lo lo

quicksort :: MV.IOVector Word64 -> Int -> Int -> IO ()
quicksort arr lo hi
  | lo >= hi  = return ()
  | otherwise = do
      p <- partition arr lo hi
      quicksort arr lo (p - 1)
      quicksort arr (p + 1) hi

-- Mergesort (top-down recursive with temp buffer)
mergesortRec :: MV.IOVector Word64 -> MV.IOVector Word64 -> Int -> Int -> IO ()
mergesortRec arr buf lo hi
  | lo >= hi  = return ()
  | otherwise = do
      let mid = lo + (hi - lo) `div` 2
      mergesortRec arr buf lo mid
      mergesortRec arr buf (mid + 1) hi
      -- copy to buf
      let copyLoop !k
            | k > hi    = return ()
            | otherwise = MV.read arr k >>= MV.write buf k >> copyLoop (k+1)
      copyLoop lo
      -- merge back
      let mergeLoop !i !j !k
            | i > mid   = if j > hi then return ()
                          else MV.read buf j >>= MV.write arr k >> mergeLoop i (j+1) (k+1)
            | j > hi    = MV.read buf i >>= MV.write arr k >> mergeLoop (i+1) j (k+1)
            | otherwise = do
                vi <- MV.read buf i
                vj <- MV.read buf j
                if vi <= vj
                  then MV.write arr k vi >> mergeLoop (i+1) j (k+1)
                  else MV.write arr k vj >> mergeLoop i (j+1) (k+1)
      mergeLoop lo (mid + 1) lo

checksum :: MV.IOVector Word64 -> IO Word64
checksum arr = do
    let n = MV.length arr
        go !i !h
          | i >= n    = return h
          | otherwise = do
              x <- MV.read arr i
              go (i+1) (h * 131 + x)
    go 0 0

main :: IO ()
main = do
    args <- getArgs
    case args of
      [algo, ns] -> do
        let n = read ns :: Int
        arr <- generateArray n 42

        t0 <- getTime Monotonic
        case algo of
          "quick" -> quicksort arr 0 (n - 1)
          "merge" -> do
            buf <- MV.new n
            mergesortRec arr buf 0 (n - 1)
          _ -> error $ "Unknown algo: " ++ algo
        t1 <- getTime Monotonic

        let ms = fromIntegral (toNanoSecs t1 - toNanoSecs t0) / 1e6 :: Double
        cs <- checksum arr
        putStrLn $ algo ++ " n=" ++ show n ++ " " ++ show ms ++ "ms checksum=" ++ show cs
      _ -> putStrLn "Usage: sort <quick|merge> <n>"

File: list-sort/rust/src/bin/sort.rs
use std::env;
use std::time::Instant;

/* LCG PRNG (Knuth) — identical across all languages */
struct Lcg(u64);

impl Lcg {
    fn new(seed: u64) -> Self { Lcg(seed) }
    fn next(&mut self) -> u64 {
        self.0 = self.0.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
        self.0
    }
}

/* Quicksort (Lomuto partition) */
fn quicksort(a: &mut [u64]) {
    if a.len() <= 1 { return; }
    let hi = a.len() - 1;
    let pivot = a[hi];
    let mut i = 0;
    for j in 0..hi {
        if a[j] <= pivot {
            a.swap(i, j);
            i += 1;
        }
    }
    a.swap(i, hi);
    let (left, right) = a.split_at_mut(i);
    quicksort(left);
    quicksort(&mut right[1..]);
}

/* Mergesort */
fn mergesort(a: &mut [u64], buf: &mut [u64]) {
    let n = a.len();
    if n <= 1 { return; }
    let mid = n / 2;
    mergesort(&mut a[..mid], &mut buf[..mid]);
    mergesort(&mut a[mid..], &mut buf[mid..]);
    // merge
    buf[..n].copy_from_slice(&a[..n]);
    let (mut i, mut j, mut k) = (0, mid, 0);
    while i < mid && j < n {
        if buf[i] <= buf[j] { a[k] = buf[i]; i += 1; }
        else                 { a[k] = buf[j]; j += 1; }
        k += 1;
    }
    while i < mid { a[k] = buf[i]; i += 1; k += 1; }
    while j < n   { a[k] = buf[j]; j += 1; k += 1; }
}

fn checksum(a: &[u64]) -> u64 {
    a.iter().fold(0u64, |h, &x| h.wrapping_mul(131).wrapping_add(x))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 3 {
        eprintln!("Usage: {} <algo: quick|merge> <n>", args[0]);
        std::process::exit(1);
    }
    let algo = &args[1];
    let n: usize = args[2].parse().unwrap();

    let mut rng = Lcg::new(42);
    let mut a: Vec<u64> = (0..n).map(|_| rng.next()).collect();

    let start = Instant::now();

    match algo.as_str() {
        "quick" => quicksort(&mut a),
        "merge" => {
            let mut buf = vec![0u64; n];
            mergesort(&mut a, &mut buf);
        }
        _ => { eprintln!("Unknown algo: {}", algo); std::process::exit(1); }
    }

    let ms = start.elapsed().as_secs_f64() * 1000.0;
    println!("{} n={} {:.1}ms checksum={}", algo, n, ms, checksum(&a));
}
